<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [3]:
import pandas as pd
import psycopg2

In [4]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
# !!!!!!DELETE!!!!!!
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies).

In [6]:
# текст запроса
query_3_1 = f'''
                select *
                from vacancies
            '''

In [7]:
# результат запроса
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_vacancies = pd.read_sql_query(query_3_1, connection)
df_vacancies.shape[0]


<ipython-input-7-372ea47881fa>:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacancies = pd.read_sql_query(query_3_1, connection)


49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers).

In [8]:
# текст запроса
query_3_2 = f'''
                select *
                from employers
            '''


In [9]:
# результат запроса
df_employers = pd.read_sql_query(query_3_2, connection)
df_employers.shape[0]
#df_employers.info()

<ipython-input-9-1d0f0e05e44d>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employers = pd.read_sql_query(query_3_2, connection)


23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [10]:
# текст запроса
query_3_3 = f'''
                select *
                from areas
            '''


In [11]:
# результат запроса
df_areas = pd.read_sql_query(query_3_3, connection)
df_areas.shape[0]
#df_areas.info()

<ipython-input-11-680cc1dace21>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_areas = pd.read_sql_query(query_3_3, connection)


1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [12]:
# текст запроса
query_3_4 = f'''
                select *
                from industries
            '''

In [13]:
# результат запроса
df_industries = pd.read_sql_query(query_3_4, connection)
df_industries.shape[0]
#df_industries.info()

<ipython-input-13-963b7b3c82c8>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_industries = pd.read_sql_query(query_3_4, connection)


294

***

# Выводы по предварительному анализу данных

В предоставленных данных видим:
* более 49 тыс. вакансий;
* 23,5 тыс. компаний-работодателей;
* 294 сферы деятельности;
* 1362 возможных городов.

Количесво вакансий превышает работодателей более чем в 2 раза.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [14]:
# текст запроса
query_4_1 = f'''
            select
                a.name area,
                count(v.id) cnt
            from vacancies v
            join areas a on a.id=v.area_id
            group by a.id
            order by count(v.id) desc
            limit 5
            '''

In [66]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1


<ipython-input-66-e0d1ca36e390>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [16]:
# текст запроса
query_4_2 = f'''
            select count(id) cnt
            from vacancies
            where salary_from is not null or salary_to  is not null
            '''


In [17]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2.cnt[0]



<ipython-input-17-1cae8679837b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [18]:
# текст запроса
query_4_3 = f'''
            select
                round(avg(salary_from)) avg_salary_from,
                round(avg(salary_to)) avg_salary_to
            from vacancies
            '''


In [19]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

<ipython-input-19-5722848d6806>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [87]:
# текст запроса
query_4_4 = f'''
            select
                schedule,
                employment,
                count(id)
            from vacancies
            group by schedule, employment
            order by count(id) desc
            limit 5
            --offset 1 limit 1
            '''

In [88]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

<ipython-input-88-f3d798c9d673>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0       Полный день     Полная занятость  35367
1  Удаленная работа     Полная занятость   7802
2     Гибкий график     Полная занятость   1593
3  Удаленная работа  Частичная занятость   1312
4    Сменный график     Полная занятость    940

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [22]:
# текст запроса
query_4_5 = f'''
            select
                experience,
                count(id)
            from vacancies
            group by experience
            order by count(id)
            '''

In [23]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

<ipython-input-23-a407e816ef92>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# Выводы по детальному анализу вакансий

Топ 5 городов по количеству вакансий:
* Москва	5333
* Санкт-Петербург	2851
* Минск	2112
* Новосибирск	2006
* Алматы	1892

Зарплата указана почти в половине вакансий. Средние значения для нижней и верхний границы составляют 71 тыс.руб. и 110 тыс.руб.

По типу трудоустройства и формату рабочего графика в вакансиях на 1-месте Полная занятость/Полный день - 71%, на 2 месте - удаленный формат при полной занятости - 16%. И в 93% это все равно полная занятость.

В больше чем половине случаев работодатели ищут сотрудников с небольшим опытом работы от 1 года до 3 лет.





# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [93]:
## текст запроса
query_5_1 = f'''
            (select
                e.name employer,
                count(v.id) cnt
            from vacancies v
            join employers e on e.id=v.employer_id
            group by e.id
            order by count(v.id) desc
            limit 1)
            union
            (select
                e.name employer,
                count(v.id) cnt
            from vacancies v
            join employers e on e.id=v.employer_id
            group by e.id
            order by count(v.id) desc
            offset 4 limit 1)
            order by 2 desc
            '''

In [95]:
## текст запроса

query_5_1 = f'''
            select
                e.name employer,
                count(v.id) cnt
            from vacancies v
            join employers e on e.id=v.employer_id
            group by e.id
            order by count(v.id) desc
            limit 5
            '''

In [96]:
# результат запроса ???
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

<ipython-input-96-66ed7a91f666>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


employer   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [97]:
# текст запроса
query_5_2 = f'''
            select
                a.name area,
                count(distinct(e.id)) cnt_employers
            from areas a
            join employers e on a.id=e.area
            left join vacancies v on a.id=v.area_id
            where v.id is null
            group by a.id
            order by count(e.id) desc
            limit 1
            '''


In [98]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

<ipython-input-98-14e6adc3ead1>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


area  cnt_employers
0  Россия            410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [103]:
# текст запроса
query_5_3 = f'''
            select
                e.name employer_name,
                count(distinct(v.area_id)) cnt_area
            from vacancies v
            join employers e on e.id=v.employer_id
            group by e.name
            order by count(v.area_id) desc
            limit 5
            --limit 1
            '''

In [104]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

<ipython-input-104-e83115aae06f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


employer_name  cnt_area
0         Яндекс       181
1     Ростелеком       152
2       Тинькофф        43
3           СБЕР        24
4  Газпром нефть        24

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [105]:
# текст запроса
query_5_4 = f'''
            select
                count(e.id)
            from employers e
            left join EMPLOYERS_INDUSTRIES ei on ei.employer_id = e.id
            where ei.industry_id is null
            '''

In [106]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

<ipython-input-106-59cd48778173>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [111]:
# текст запроса
query_5_5 = f'''
            select
                e.name
            from employers e
            left join EMPLOYERS_INDUSTRIES ei on ei.employer_id = e.id
            group by e.id
            having count(ei.industry_id) = 4
            order by e.name
            offset 2 limit 1
            '''

In [112]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

<ipython-input-112-55a59fdce066>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [113]:
# текст запроса
query_5_6 = f'''
            select
                count(e.name)
            from employers e
            join EMPLOYERS_INDUSTRIES ei on ei.employer_id = e.id
            join industries i on ei.industry_id=i.id
            where i.name ='Разработка программного обеспечения'
            '''

In [114]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

<ipython-input-114-b8cbf15ac659>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8).

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API.

In [115]:
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
cities = tuple(pd.read_html(url)[1]['Город'])


In [117]:
#cities

In [118]:
# текст запроса
query_5_7 = f'''
              select
                  a.name,
                  count(v.id) cnt
              from vacancies v
              join areas a on v.area_id = a.id
              join employers e on e.id = v.employer_id
              WHERE e.name = 'Яндекс' and a.name in {cities}
              group by a.name

              union all

              select
                  'Total',
                  sum(cnt)
              from (select
                      a.name,
                      count(v.id) cnt
                  from vacancies v
                  join areas a on v.area_id = a.id
                  join employers e on e.id = v.employer_id
                  WHERE e.name = 'Яндекс' and a.name in {cities}
                  group by a.name) as total
              order by 2
            '''

In [119]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

<ipython-input-119-0bbb2c17c37c>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7            Самара   26.0
8               Уфа   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

# Выводы по анализу работодателей

Топ-5 Работодателей по количеству размещенных вакансий:
* Яндекс
* Ростелеком
* Тинькофф
* СБЕР
* Газпром нефть

Эти же компании являются лидерами по количеству регионов, в которых размещены вакансии.

Ожидаемо крупные компании и максимально широко представлении на территории России.

У Яндекс в 16 городах-миллионках размещены 485 вакансиий (25% от общего числа всех вакансий Яндекса)

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [43]:
# текст запроса
query_6_1 = f'''
            select
                count(id)
            from vacancies
            where lower(name) like '%data%' or lower(name) like '%данн%'
            '''


In [44]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

<ipython-input-44-e8ca8a3733ce>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.


In [53]:
# текст запроса
query_6_2 = f'''
            select
                count(id)
            from vacancies
            where ((name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or name ilike '%machine learning%'
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%')
                    or (name not like '%HTML%' and name like '%ML%'))
                    and
                    (name ilike '%junior%' or experience = 'Нет опыта' or employment = 'Стажировка')
            '''


In [51]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

<ipython-input-51-b7a554361891>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [54]:
# текст запроса
query_6_3 = f'''
            select
                count(id)
            from vacancies
            where ((name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or name ilike '%machine learning%'
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%')
                    or (name not like '%HTML%' and name like '%ML%'))
                    and
                    (key_skills ilike '%SQL%' or key_skills ilike '%postgres%')
            '''


In [55]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

<ipython-input-55-8a4199a81e3b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [56]:
# текст запроса
query_6_4 = f'''
            select
                count(id)
            from vacancies
            where ((name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or name ilike '%machine learning%'
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%')
                    or (name not like '%HTML%' and name like '%ML%'))
                    and
                    (key_skills ilike '%Python%')
            '''


In [57]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

<ipython-input-57-95116ecfece2>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [60]:
# текст запроса
query_6_5 = f'''
            select
                round(AVG(length(key_skills) - length(replace(key_skills, CHR(9),'')) + 1),2)
            from vacancies
            where ((name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or name ilike '%machine learning%'
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%')
                    or (name not like '%HTML%' and name like '%ML%'))
                    and (key_skills IS NOT NULL)
            '''


In [61]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

<ipython-input-61-1cefb534f424>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*).

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [64]:
# текст запроса
query_6_6 = f'''
            select
                experience,
                round(AVG((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from))/2),0) mean_salary

            from vacancies
            where ((name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or name ilike '%machine learning%'
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%')
                    or (name not like '%HTML%' and name like '%ML%'))
                    and (salary_from is not NULL or salary_to is not NULL)
            group by experience
            '''


In [65]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

<ipython-input-65-12f33d1ad98b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


experience  mean_salary
0           Нет опыта      74643.0
1  От 1 года до 3 лет     139675.0
2       От 3 до 6 лет     243115.0

***

# Выводы по предметному анализу

- 1771 вакансий имеют отношения к данным;
- 51 вакансий подходит для начинающих Датасайнтистов;
- 201 вакансий для ДатаСайнтистов с навыками SQL и postgres;
- 351 вакансий для ДатаСайнтистов с навыками Python.

В среднем в вакансиях ДатаСайнтистов указывают 6 ключевых навков.

ДатаСайнтисты в среднем могут расчитыват на зарплату:
* без опыта - 74 тыс.руб.
* от 1 года до 3 лет - 140 тыс.руб.
* с опытом от 3-6 лет - 243 тыс.руб.


# Общий вывод по проекту

Проведен анализ базы данных вакансий.
На первоначальном этапе знакомство было определено, что в данных представлено 49 тыс. вакансий.
В Москве открыто наибольшее количество вакансий. Зарплату указывают в половину случаях. Полная занятость требуется в большинстве случаях.
Выявили наиболее крупных работадалей по количеству вакансий и регионам присутвия.
Исследовали на примере интересующей нас вакансии ДатаСайнтиста, количество размещенных вакансий, требования к ключевым навыкам, а так же зарплатные ожидания в зависиомсти от опыта.
